org.lflang/src/org/lflang/federated/extensions/CExtensionUtils.java

```diff
@@ -246,13 +246,25 @@ public static void handleCompileDefinitions(
         federate.targetConfig.compileDefinitions.put("FEDERATED_"+federate.targetConfig.coordination.toString().toUpperCase(), "");
         federate.targetConfig.compileDefinitions.put("NUMBER_OF_FEDERATES", String.valueOf(numOfFederates));
         federate.targetConfig.compileDefinitions.put("EXECUTABLE_PREAMBLE", "");
-        federate.targetConfig.compileDefinitions.put("WORKERS_NEEDED_FOR_FEDERATE", String.valueOf(federate.networkMessageActions.size()));
+        federate.targetConfig.compileDefinitions.put("WORKERS_NEEDED_FOR_FEDERATE", String.valueOf(minThreadsToHandleInputPorts(federate)));
 
         handleAdvanceMessageInterval(federate);
 
         initializeClockSynchronization(federate, federationRTIProperties);
     }
 
+    /**
+     * The number of threads needs to be at least one larger than the input ports
+     * to allow the federate to wait on all input ports while allowing an additional
+     * worker thread to process incoming messages.
+     * @return The minimum number of threads needed.
+     */
+    public static int minThreadsToHandleInputPorts(FederateInstance federate) {
```

Comment:
I'm amazed that you were able to pinpoint this problem and implement a fix! Well done! That said, this shows again that the approach is exceptionally brittle and won't scale well. We need a better solution...